### PR TITLE
Improve documentation on Registers and Qubits

### LINKED ISSUE #1469: Better documentation on Registers and Qubits
### URL, if applicable

Probably https://docs.quantum.ibm.com/build/circuit-construction

### Describe the new content you are requesting.

This is user feedback from Qiskit Slack. A user asked this SO question: https://quantumcomputing.stackexchange.com/questions/38565/unable-to-extract-the-quantum-registers-information-from-qiskit-quantum-circuit

> slack user 2: Once you have your Qubit isolated you can retrieve the index by doing:
qc.qubits.index(your_qubit)
where qc is your QuantumCircuit object

>> slack user 1: Thank you very much! I have been searching all over the internet for the answer for the past 2 days

> Yes I find the documentation to be not super clear on how Registers and Qubits are handled

### If this new content request is accepted, do you want to write the content?

no

### PR DESCRIPTION
**Pull request by shraddha-aangiras**
Based on user feedback from the Qiskit Slack, I added more explanations of commonly used methods and attributes of circuits. In particular:
- explained how to extract information from circuits
- added information on creating and managing registers
- added information measuring circuits 
fixes #1469

### CODE REVIEW

File: docs/build/circuit-construction.ipynb [hunk 1]
```diff
@@ -349,6 +403,105 @@
...
+    "qc.measure(0, 1)\n",
+    "qc.draw(\"mpl\", cregbundle=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "79ba9f07",
+   "metadata": {},
+   "source": [
+    "* `QuantumCircuit.measure_all` : takes no argument and can be used for quantum circuits without pre-defined classical bits. It will create classical wires and store measurement results in order. It will also introduce a barrier before the measurement."
```
Comment (javabster) [inline docs/build/circuit-construction.ipynb:445]: for these bullet points about the different types of measurement I think it would be better to group them all together before the code examples.

Comment (javabster) [inline docs/build/circuit-construction.ipynb:445]: also for this line in particular can you rephrase "store measurement results in order". its not clear to me here what you mean by "in order".

File: docs/build/circuit-construction.ipynb [hunk 2]
```diff
@@ -50,6 +50,74 @@
...
+    "combined_circ = QuantumCircuit(qr1, qr2, cr1) # Create a quantum circuit with 2 QuantumRegisters and 1 ClassicalRegister\n",
+    "combined_circ.qubits"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6e7b212f",
+   "metadata": {},
+   "source": [
+    "The index and register of a qubit can be obtained using the circuit's [`find_bit`](/api/qiskit/qiskit.circuit.QuantumCircuit#qiskit.circuit.QuantumCircuit.find_bit) method and its attributes. "
```
Comment (javabster) [inline docs/build/circuit-construction.ipynb:96]: I believe in the original issue attached to this PR it was mentioned that a user was especially interested in the `qc.qubits.index(your_qubit)` method of getting information. could you include something about this as well as the `find_bit` method?

**Review by javabster — COMMENTED** [review]
Thanks so much, this is a lovely contribution! I just left a couple of minor comments but other than that its looking really good

**Comment by frankharkins** [comment]
@shraddha-aangiras we've recently edited the history on `main`, can you please reset your branch using the following commands.

```sh
# Make sure you're on your branch
git switch improvedocs-issue1469
# Overwrite your branch with our history
git reset --hard origin/main
# re-apply your commits on top of the new history
git cherry-pick c370ad180de77c0b816dc025dd719806112f753a  361ac3ffb84190f25930a9c6a94819f4eb058c9f
# Check everything looks correct
git diff main
# If all good, force-push to your branch
git push --force
```

Let me know if you have any questions. Sorry for the inconvenience.

**Comment by shraddha-aangiras** [comment]
@frankharkins No problem, will get that along with the suggested changes (thanks @javabster!) fixed by tomorrow

File: docs/build/circuit-construction.ipynb (hunk 2 above)
Comment (shraddha-aangiras) [reply docs/build/circuit-construction.ipynb]: I didn't include that due to [this comment in a previous PR ](https://github.com/Qiskit/documentation/pull/1587#discussion_r1649463291) which says it's not recommended due to the significant performance overhead. Please let me know whether to include it or not in reference to this, thanks!

**Comment by shraddha-aangiras** [comment]
@frankharkins can you let me know if I did it right? Thanks!

File: docs/guides/construct-circuits.ipynb [hunk 3]
```diff
@@ -349,6 +403,93 @@
     "qc_a.decompose().draw(\"mpl\")"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "87f1e053",
+   "metadata": {},
+   "source": [
+    "Qubits can be measured using three methods: [`measure`](/api/qiskit/qiskit.circuit.QuantumCircuit#qiskit.circuit.QuantumCircuit.measure), [`measure_all`](/api/qiskit/qiskit.circuit.QuantumCircuit#measure_all) and [`measure_active`](/api/qiskit/qiskit.circuit.QuantumCircuit#measure_active). To learn how to visualize measured results, see the [Visualize results](/run/visualize-results) page.\n",
```
Comment (frankharkins) [inline docs/guides/construct-circuits.ipynb:411]: This has changed since the PR was made

```suggestion
    "Qubits can be measured using three methods: [`measure`](/api/qiskit/qiskit.circuit.QuantumCircuit#qiskit.circuit.QuantumCircuit.measure), [`measure_all`](/api/qiskit/qiskit.circuit.QuantumCircuit#measure_all) and [`measure_active`](/api/qiskit/qiskit.circuit.QuantumCircuit#measure_active). To learn how to visualize measured results, see the [Visualize results](/guides/visualize-results) page.\n",
```

File: docs/guides/construct-circuits.ipynb [hunk 4]
```diff
@@ -85,25 +153,23 @@
    "id": "17a82d8a-b717-44b8-b3f8-ce89e2588261",
    "metadata": {},
    "source": [
-    "The easiest way to view this information is through the [`draw`](/api/qiskit/qiskit.circuit.QuantumCircuit#draw) method, which returns a visualization of a circuit. See [Visualize circuits](./visualize-circuits) for different ways of displaying quantum circuits."
+    "The easiest way to view this information is through the [`draw`](/api/qiskit/qiskit.circuit.QuantumCircuit#draw) method, which returns a visualization of a circuit. See [Visualize circuits](/build/circuit-visualization) for different ways of displaying quantum circuits."
```
Comment (frankharkins) [inline docs/guides/construct-circuits.ipynb:156]:
```suggestion
    "The easiest way to view this information is through the [`draw`](/api/qiskit/qiskit.circuit.QuantumCircuit#draw) method, which returns a visualization of a circuit. See [Visualize circuits](/guides/visualize-circuits) for different ways of displaying quantum circuits."
```

File: docs/build/circuit-construction.ipynb (hunk 2 above)
Comment (frankharkins) [reply docs/build/circuit-construction.ipynb]: I agree with @shraddha-aangiras. The two approaches have the same output, but `find_bit` is more efficient.

**Review by frankharkins — COMMENTED** [review]
@shraddha-aangiras looks great! I have a few suggestions.

Also, sorry about my instructions, but glad you worked it out. (For others reading: `git reset --hard origin/main` only works if "origin" refers to this repo, which is probably not the case for external contributors.)

**Review by kaelynj — COMMENTED** [review]
This is a great contribution thank you! This satisfies what I was thinking of when I opened #1601

The only thing I'd ask before approving is if you could run the notebook linter to make sure all of the formatting is consistent. You'll just need to run
`tox -e fix -- docs/**/*.ipynb` if you have tox installed.

File: docs/guides/construct-circuits.ipynb [hunk 5]
```diff
@@ -349,6 +403,93 @@
     "qc_a.decompose().draw(\"mpl\")"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "87f1e053",
+   "metadata": {},
+   "source": [
+    "Qubits can be measured using three methods: [`measure`](/api/qiskit/qiskit.circuit.QuantumCircuit#qiskit.circuit.QuantumCircuit.measure), [`measure_all`](/api/qiskit/qiskit.circuit.QuantumCircuit#measure_all) and [`measure_active`](/api/qiskit/qiskit.circuit.QuantumCircuit#measure_active). To learn how to visualize measured results, see the [Visualize results](/guides/visualize-results) page.\n",
```
Comment (beckykd) [inline docs/guides/construct-circuits.ipynb:411]: Can we add a header here and a bit of an intro?  I'm not sure if it should be level 2 or 3.  Maybe something like this (I'm not even close to a quantum physicist, so it should be fixed up to be accurate): 

<span id="measure-qubits"></span>
## Measure qubits

Measurements can be used to get the final results from the quantum register to the classical register.  Note that if you are submitting circuits to a Sampler primitive, measurements are required.  However, circuits submitted to an Estimator primitive cannot contain measurements.

File: docs/guides/construct-circuits.ipynb [hunk 6]
```diff
@@ -349,6 +403,93 @@
...
+    "\n",
+    "1. `QuantumCircuit.measure` : measures each qubit in the first argument into the classical bit given as the second argument. This method allows full control on where the measurement result is stored.\n",
```
Comment (beckykd) [inline docs/guides/construct-circuits.ipynb:413]:
```suggestion
    "1. `QuantumCircuit.measure` : measures each qubit in the first argument onto the classical bit given as the second argument. This method allows full control over where the measurement result is stored.\n",
```

File: docs/guides/construct-circuits.ipynb [hunk 7]
```diff
@@ -349,6 +403,93 @@
...
+    "\n",
+    "2. `QuantumCircuit.measure_all` : takes no argument and can be used for quantum circuits without pre-defined classical bits. It will create classical wires and store measurement results in order (measurement of a qubit $q_i$ is stored in cbit $meas_i$). It will also introduce a barrier before the measurement.\n",
```
Comment (beckykd) [inline docs/guides/construct-circuits.ipynb:415]:
```suggestion
    "2. `QuantumCircuit.measure_all` : takes no argument and can be used for quantum circuits without pre-defined classical bits. It creates classical wires and stores measurement results in order. For example, measurement of  qubit $q_i$ is stored in cbit $meas_i$). It also adds a barrier before the measurement.\n",
```

File: docs/guides/construct-circuits.ipynb [hunk 8]
```diff
@@ -349,6 +403,93 @@
...
+    "\n",
+    "3. `QuantumCircuit.measure_active` : similar to `measure_all`, but measures only qubits that have operations"
```
Comment (beckykd) [inline docs/guides/construct-circuits.ipynb:417]:
```suggestion
    "3. `QuantumCircuit.measure_active` : similar to `measure_all`, but measures only qubits that have operations."
```

**Review by beckykd — COMMENTED** [review]
This is great!  Thank you so much for your contribution!

File: docs/guides/construct-circuits.ipynb (hunk 5 above)
Comment (shraddha-aangiras) [reply docs/guides/construct-circuits.ipynb:415]: Added, thanks!

File: docs/guides/construct-circuits.ipynb [hunk 9]
```diff
@@ -23,7 +23,7 @@
    "source": [
     "## What is a quantum circuit?\n",
     "\n",
-    "A simple quantum circuit is a collection of qubits and a list of instructions that act on those qubits. To demonstrate, the following cell creates a new circuit with two new qubits, then displays the circuit's [`qubits`](/api/qiskit/qiskit.circuit.QuantumCircuit#qubits) attribute."
+    "A simple quantum circuit is a collection of qubits and a list of instructions that act on those qubits. To demonstrate, the following cell creates a new circuit with two new qubits, then displays the circuit's [`qubits`](/api/qiskit/qiskit.circuit.QuantumCircuit#qubits) attribute, which is a list of [`Qubits`](/api/qiskit/circuit#qiskit.circuit.Qubit) in order."
```
[the remark this answers is not in the document]
Comment (shraddha-aangiras) [reply docs/guides/construct-circuits.ipynb]: Clarified this in the latest commit, thank you!

**Comment by shraddha-aangiras** [comment]
Thanks @beckykd @kaelynj, made the changes you suggested :)

File: docs/guides/construct-circuits.ipynb [hunk 10]
```diff
@@ -349,6 +403,97 @@
     "qc_a.decompose().draw(\"mpl\")"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "87f1e053",
+   "metadata": {},
+   "source": [
+    "## Measure qubits\n",
```
Comment (beckykd) [inline docs/guides/construct-circuits.ipynb:411]:
```suggestion
"<span id=\"measure-qubits\"></span>\n",
    "## Measure qubits\n",
```

File: docs/guides/construct-circuits.ipynb [hunk 11]
```diff
@@ -23,7 +23,7 @@
    "source": [
```
Comment (kaelynj) [inline docs/guides/construct-circuits.ipynb:23]: <p>I would suggest changing this first couple sentences to:</p><p> </p><p>Measurements are used to sample the states of individual qubits and transfer the results to a classical register. Note that if you are submitting circuits to a <a href="https://docs.quantum.ibm.com/guides/primitives#sampler" target="_blank">Sampler primitive</a>, measurements are required. However, circuits submitted to an <a href="https://docs.quantum.ibm.com/guides/primitives#estimator" target="_blank">Estimator primitive</a> <em>must not</em> contain measurements.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/1601/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>
